[5단계 미션]  심혁 미션 제출합니다.

Johncakes opened the pull request:
안녕하세요 최혜령 리뷰어님! 3기 프론트엔드 심혁입니다. 🙌 
미션과 리뷰를 통해 많은 것을 배우고 있으며, 리뷰어님들의 소중한 의견 덕분에 성장하고 있는 것 같습니다.
앞으로도 잘 부탁드리겠습니다. 감사합니다!

# 이번 미션 변경점 및 느낀점
## API 추가
이번 미션에는 로딩 상태와 에러 상태를 고려하지 않는 조건이어서 try...catch는 사용 안 했습니다.

초기 구현 시에는 App.jsx와 AddRestaurantModal 각각에 데이터를 조회하고 저장하는 로직을 분산하여 작성했습니다. 특히 App.jsx에서는 데이터를 가져오는 작업을, 모달 컴포넌트에서는 데이터를 업로드하는 작업을 위해 두 파일 모두에 useEffect를 사용했습니다.

이 방식은 기능상 문제없이 동작했지만, 불필요한 중복 코드와 두 번의 useEffect 사용이 비효율적이라고 판단되어 개선을 고민했습니다.

이러한 고민 끝에, 저번 미션 리뷰에서 배운 커스텀 훅을 활용하여 데이터 관련 로직을 통합하기로 결정했습니다. 그 결과, 데이터를 조회하는 함수와 데이터를 저장한 후 재조회하는 함수를 하나의 커스텀 훅으로 구현하게 되었습니다.
  
## 접근성 추가
저번 미션에서 다른 스터디 인원들의 접근성 관련 코드를 보고 저도 해당 기능을 추가하게 되었습니다. 사실 개인 프로젝트를 진행할 때는 접근성을 깊이 고려해 본 적이 없었기에 이번 경험이 새롭게 다가왔습니다.

'어떻게' 그리고 '무엇을' 변경해야 사용자 접근성이 향상되는지에 대해 고민해야 한다는 것을 이번 미션을 통해 중요한 깨달음으로 얻어 가는것 같습니다. 하지만 아직 경험이 많지 않다보니 놓치는 부분도 많다는 것을 알게 되었습니다.

## useMemo 추가
- FilteredRestaurants 계산 로직에 useMemo를 적용하여, 필터링 조건이 변경되지 않는 한 이전 계산 결과를 재사용하도록 했습니다.

## Data폴더 삭제
- 데이터는 이제 서버로부터 받아오니 해당 프로젝트에서 삭제했습니다.

--- 

## 질문사항 🤔 
- 데이터를 서버에 저장하는 postRestaurant 함수를 구현하면서, 저장 후 데이터 목록을 즉시 최신 상태로 갱신하기 위해 해당 함수 내에서 fetchRestaurants 함수를 호출하도록 했습니다. 데이터를 post하는 로직 안에 데이터를 fetch하는 로직을 포함하는 이 방식이 커스텀 훅의 역할 분리나 설계 관점에서 괜찮은지 리뷰어님의 의견을 여쭙고 싶습니다.
```
  const postRestaurant = async (restaurant) => {
    await addRestaurant(restaurant);
    await fetchRestaurants();
  };
``` 
## 리뷰 요청 사항
- API 활용 : 구현된 API 호출 및 데이터 처리 로직이 적절한지, 그리고 더 효율적이거나 모범적인 접근 방식은 없는지 검토 부탁드립니다!
- 접근성 : 웹 접근성 측면에서 놓친 부분이나 개선이 필요한 요소가 있는지 검토해 주시면 감사하겠습니다!

## 프로젝트 구성
<img width="1831" height="1194" alt="Step 5" src="https://github.com/user-attachments/assets/4c359eb2-f361-4181-90fe-1c185bcfcb58" />


## 실행 영상
https://github.com/user-attachments/assets/f1f1b083-fca4-402b-bc29-e550efcd161c

Johncakes committed:
feat: Header component 추가

Johncakes committed:
feat: App.css styling 추가

Johncakes committed:
feat: RestaurantCategoryFilter component 추가

Johncakes committed:
feat: RestaurantList component 추가

Johncakes committed:
feat: RestaurantDetailModal component 추가

Johncakes committed:
feat: AddRestaurantModal component 추가

Johncakes committed:
chore: 이미지 파일 assets으로 이동

Johncakes committed:
feat: 모든 component App.jsx에 추가

Johncakes committed:
feat: 모든 component App.jsx에 추가

Johncakes committed:
style: prettier 적용

Johncakes committed:
docs: 요구사항 및 느낀점 추가

Johncakes committed:
docs: 요구사항 및 느낀점 추가

Johncakes committed:
feat: restaurants data 추가

Johncakes committed:
refactor: category를 constant 파일로 변경

Johncakes committed:
feat: getCategoryIcon util 함수 추가

Johncakes committed:
feat: getCategoryIcon util 함수 추가

Johncakes committed:
feat: getFilteredRestaurant util 함수 추가

Johncakes committed:
feat: getFilteredRestaurant util 함수 추가

Johncakes committed:
refactor: RestaurnantList 재구성

- FilteredRestaurants로 받는 정보로 map하여 RestaurantListItem으로 표시
- FilteredRestaurants에 icon을 받을수 있게 getCategoryIcon의 결과값을 props로 보냄.

Johncakes committed:
refactor: RestaurantListItem 컴포넌트화

- 받는 props로 restaurant 이름, icon, description을 설정

Johncakes committed:
refactor: RestaurantCategory options map으로 재구성

- category를 constants에서 받아와 map을 사용

Johncakes committed:
feat: useState 사용과 컴포넌트 props 추가

Johncakes committed:
feat: useState 사용과 컴포넌트 props 추가

Johncakes committed:
docs: step2 docs 수정

Johncakes committed:
docs: step2 docs 수정

Johncakes committed:
chore: remove comment in app

Johncakes committed:
chore: remove comment in app

Johncakes committed:
refactor: category state rename

state의 category와 constants의category가 같아서 네이밍 수정

Johncakes committed:
docs: README.md 에서 development-log.md로 변경

Johncakes committed:
docs: README.md 에서 development-log.md로 변경

Johncakes committed:
fix: 중식 카테고리 아이콘이 안나오는 오류 수정

Johncakes committed:
fix: 중식 카테고리 아이콘이 안나오는 오류 수정

Johncakes committed:
chore: 네이밍 일관성있게 수정

Johncakes committed:
refactor: return 값 읽기 쉽게 수정

Johncakes committed:
refactor: return 값 읽기 쉽게 수정

Johncakes committed:
refactor: getCategoryIcon 함수 유지보수 좋은 코드로 수정

Johncakes committed:
refactor: getCategoryIcon 함수 유지보수 좋은 코드로 수정

Johncakes committed:
chore: EOF 없는 파일 수정

Johncakes committed:
refactor: image 관리 파일 추가

Johncakes committed:
refactor: 파일 이름 및 경로 수정

Johncakes committed:
refactor: 파일 이름 및 경로 수정

Johncakes committed:
refactor: self-closing tag으로 수정

Johncakes committed:
refactor: Object.freeze() for categories const

Johncakes committed:
refactor: rename 'Style' to 'Styles'

Johncakes committed:
refactor: change onInput to onChange

Johncakes committed:
Merge branch 'johncakes' into step2

Johncakes committed:
Merge branch 'step2' of https://github.com/Johncakes/self-paced-react into step2

Johncakes committed:
refactor: 카테고리 선택 map함수로 구현

Johncakes committed:
feat : Modal 여닫는 기능 추가

feat : Modal 여닫는 기능 추가
- 각 Modal의 useState을 만들어서 하나씩 State값 변경할 수 있게 만듬
- Header과 RestaurantListItem에 state 변경하는 기능 추가
- 각 Modal에 state를 false로 변경하는 기능 추가

Johncakes committed:
refactor : state관리 변경

Johncakes committed:
feat: 선택된 restaurant 정보를 Modal로 넘기는 기능 추가

feat: 선택된 restaurant 정보를 Modal로 넘기는 기능 추가
- selectedRestaurant state를 만들어서 RestaurantDetail modal에 사용할수 있게 구현

Johncakes committed:
feat: Modal.jsx 추가

Johncakes committed:
refactor: Modal.jsx 사용해서 재구성

Johncakes committed:
feat: 식당 추가 기능 구현

Johncakes committed:
chore: removed development-log.md

- 같은 내용을 PR에 작성을 해서 불필요하다고 느낌.

Johncakes committed:
fix: modal 여닫는 함수 수정

- Modal에 handleClose을 보내서 해당 Modal을 닫을수 있게 수정

Johncakes committed:
fix: directory casing 문제 해결

Johncakes committed:
refactor: custom Hook으로 modal 관리

Johncakes committed:
refactor: form 구조 분해 할당으로 변경

Johncakes committed:
chore: rename to fit camelCasing

Johncakes committed:
chore: rename modal handlers

Johncakes committed:
refactor : destructuring으로 useModal 사용하는 변수들 수정

Johncakes committed:
refactor: state 네이밍 수정, 불필요한 return 값 삭제

Johncakes committed:
Merge branch 'johncakes' into step4

Johncakes committed:
feat: API 기능 추가

Johncakes committed:
feat: increased accessibility

Johncakes committed:
chore: delete static data

Johncakes committed:
refactor: use 'autofocus' for modal focus

HyeryongChoi commented on src/components/modal/AddRestaurantModal.jsx:
```diff
@@ -94,6 +92,7 @@ export default function AddRestaurantModal({ closeModal }) {
           <button
             type="submit"
             className={`${styles.button} ${styles["button--primary"]} text-caption`}
+            aria-label="추가하기"
```
버튼에 이미 `추가하기`텍스트가 있어서 중복 라벨링은 불필요해 보입니다~

HyeryongChoi commented on src/components/modal/RestaurantDetailModal.jsx:
```diff
@@ -13,6 +13,8 @@ export default function RestaurantDetailModal({ restaurantInfo, closeModal }) {
         <button
           className={`${style.button} ${style["button--primary"]} text-caption`}
           onClick={closeModal}
+          aria-label="닫기"
```
여기도요~

HyeryongChoi commented on src/api/restaurants.js:
```diff
@@ -0,0 +1,17 @@
+  return response.json();
+};
+
+export const addRestaurant = async (restaurant) => {
```
참고용 코멘트입니다 🙂

`getRestaurants`는 GET요청을 기반으로 한 네이밍인데, `addRestaurant`는 POST요청임에도 `add`라는 도메인 명사로 되어 있어 일관성이 부족하고 조금 혼동될 수 있을 것 같아요. 한 가지 기준으로 통일하면 좋을 것 같습니다.
하지만 꼭 바꿔야 하는 건 아니고, 취향 차이일 수 있으니 참고만 해주세요.

Johncakes replied:
POST요청인 것을 생각하면 네이밍을 변경하는 게 맞다고 생각합니다. 해당 내용 수정하겠습니다!

HyeryongChoi commented on src/components/modal/AddRestaurantModal.jsx:
```diff
@@ -3,10 +3,7 @@ import styles from "./Modal.module.css";
-import restaurants from "../../data/restaurant";
-
-export default function AddRestaurantModal({ closeModal }) {
+export default function AddRestaurantModal({ closeModal, postRestaurant }) {
```
모달에서 받는 prop이름이 API 함수와 동일한 `postRestaurant`이다 보니, 모달이 직접 API 호출을 책임지는 것처럼 보입니다.

1. 모달 내부에서 훅을 직접 사용하거나
2. 외부에서 함수를 주입하되 `onSubmit~`, `onAddRestaurant`처럼 UI 관점 명칭으로 추상화하는

개선방안을 고려해볼 수 있을 것 같아요.

현재 구조에서는 UI 컴포넌트가 데이터 계층의 구체적인 명세(어떤 함수를 호출해야 하는지)에 직접 의존하기 때문에 결합도가 높아보입니다.

결과적으로 API 이름이나 구현이 바뀌면 모달 코드도 함께 수정해야 하죠. 관심사를 분리하려면, 모달은 "음식점 추가 요청이 들어오면 콜백을 호출한다"정도만 알고, 실제 API 호출 로직은 부모나 훅에서 캡슐화하는 편이 더 안정적이라고 생각합니다.

Johncakes replied:
props의 네이밍이 조금 부적절한 것 같네요. 의견 주신대로 수정을 해보았는데, 의도하신 데로 잘 수정했는지 확신이 들지 않습니다. 😓

HyeryongChoi commented on src/hooks/useRestaurants.js:
```diff
@@ -0,0 +1,22 @@
+    setRestaurants(data);
+  };
+
+  const postRestaurant = async (restaurant) => {
```
`postRestaurant` 함수 안에서 `fetchRestaurant`를 직접 호출하는 방식은 동작은 명확하지만, 관심사 분리 측면에서는 아쉬운 부분이 있습니다.

현재 방식에서는 함수가 등록뿐만 아니라 데이터 갱신까지 책임지므로, 변경이 발생할 때 수정해야 하는 지점이 늘어납니다.

등록과 데이터 갱신을 분리했을 때 어떤 장점이 있을지 한번 고민해보시면 좋을 것 같아요!

Johncakes replied:
분리했을 때 결합도 낮아질 것 같습니다. 굳이 데이터를 불러올 필요가 없는 경우 분리했을 때 불필요한 코드도 제거할 수 있을 것 같습니다. 해당 내용도 수정하겠습니다!

HyeryongChoi reviewed (changes requested):
안녕하세요, 심혁님 5단계 미션 진행하느라 고생 많으셨습니다👏

이번 미션을 통해 접근성에 대해 새롭게 고민해보셨다니 좋네요.
질문 사항에 대한 코멘트도 함께 남겨두었으니 확인해주시면 좋겠습니다.
앞으로도 놓친 부분을 하나씩 점검하며 개선해 나가면 큰 도움이 될 거예요👍

접근성 관련 참고 자료 두 가지도 공유드립니다.
- [WAI-ARIA Dialog 가이드](https://aoa.gitbook.io/skymimo/aoa-2018/2018-aria/dialog)
- [네이버 웹 접근성 가이드](https://accessibility.naver.com/accessibility)

Johncakes committed:
refactor: update useRestaurant to seperate fetch and get method

- rename addRestaurant to postRestaurant in restaurant.js
- change props from postRestaurant to onAddRestaurant in
AddRestaurantModal.jsx

Johncakes reviewed (commented):
리뷰 확인하고 반영했습니다!
의견 주신 대로 적용을 한번 해보았지만, 의도대로 잘 적용하였는지 잘 모르겠습니다. 

## 수정하면서 고민한 사항 및 질문 🤔 

`AddRestaurantModal.jsx`에 postRestaurant를 onAddRestaurant으로 바꿨습니다. 이 과정에서 `useRestaurant.js`의 훅을 수정하면서 `onAddRestaurant`에 `addRestaurants` 와 `fetchRestaurants`를 불러오는 함수를 만들었습니다. 이 방식으로 하면 `post`랑 `get`를 분리할 수 있다고 생각했습니다. 

>현재 구조에서는 UI 컴포넌트가 데이터 계층의 구체적인 명세(어떤 함수를 호출해야 하는지)에 직접 의존하기 때문에 결합도가 높아보입니다.

이 부분에서 조금 고민하게게되었습니다. 수정한한코드에 `AddRestaurantModal.jsx`의 `OnAddRestaurant`에 아직 `restaurantInfo`를 넘기기 때문에 의견 주신 코멘트를 잘 적용하지 못했다는 생각이 들었습니다.
 
UI와 데이터를 완전히 분리하기 위해 `useRestaurants.js` 안에 `restaurantInfo`를 저장할 수 있는 `state`를 만들어 보았지만, `setState `업데이트를 `AddRestaurantModal.jsx`의 `handleSumbit`안에 넣으면 `state`를 업데이트하기 전에 `onAddRestaurant`함수가 호출되기 때문에 원하는 결과를 못 얻는 것을 확인할 수 있었습니다. 
또한 이 방식으로 하는 게 오히려 코드의 가독성이 떨어진다고 느껴서 현재 수정된 방식으로 다시 바꿨습니다.

현재 제가 수정한 코드가 리뷰어님의 의도와 맞는지에 대해서 궁금합니다. 또한 제가 놓친 부분이 있으면 알려주셨으면 감사하겠습니다!

HyeryongChoi reviewed (approved):
> 수정하면서 고민한 사항 및 질문 🤔
AddRestaurantModal.jsx에 postRestaurant를 onAddRestaurant으로 바꿨습니다. 이 과정에서 useRestaurant.js의 훅을 수정하면서 onAddRestaurant에 addRestaurants 와 fetchRestaurants를 불러오는 함수를 만들었습니다. 이 방식으로 하면 post랑 get를 분리할 수 있다고 생각했습니다.
현재 구조에서는 UI 컴포넌트가 데이터 계층의 구체적인 명세(어떤 함수를 호출해야 하는지)에 직접 의존하기 때문에 결합도가 높아보입니다.
이 부분에서 조금 고민하게게되었습니다. 수정한한코드에 AddRestaurantModal.jsx의 OnAddRestaurant에 아직 restaurantInfo를 넘기기 때문에 의견 주신 코멘트를 잘 적용하지 못했다는 생각이 들었습니다.

`onAddRestaurant`에 `restaurantInfo`를 인자로 넘기는 방식은 문제가 없어요. UI 컴포넌트가 입력값을 수집하고, 그 값을 부모나 훅에서 정의한 콜백에 전달하는 것은 리액트에서 자연스러운 책임 분리 방식이에요. 제가 말씀드렸던 “UI가 데이터 계층의 구체적인 명세에 의존한다”는 건, 모달이 `postRestaurant` 같은 API 함수 자체를 직접 import해서 호출하는 상황을 말했던 것이고, 지금처럼 추상화된 콜백만 호출하는 구조에서는 결합도가 충분히 낮아진 상태입니다.

즉, 모달은 “입력된 음식점 정보를 전달한다”까지만 알고 있고, 실제로 그 정보를 이용해 API 요청을 보내거나 리스트를 갱신하는 로직은 상위에서 캡슐화되어 있기 때문에 의도대로 관심사가 분리된 구조라고 볼 수 있어요 👍

앞으로도 이런 구조적인 부분을 꾸준히 고민해보시면 도움이 될 거예요. 이번 미션도 고생 많으셨습니다 😊
